From 429eff12541cc0779c381f84257c8860ece25b12 Mon Sep 17 00:00:00 2001
From: Stefan Liebler <stli@linux.ibm.com>
Date: Wed, 11 Dec 2019 15:09:29 +0100
Subject: [PATCH 21/28] S390: Use convert-to-fixed instruction for llrint
 functions.

If compiled with z196 zarch support, the convert-to-fixed instruction
is used to implement llrint, llrintf, llrintl.
Otherwise the common-code implementation is used.

(cherry picked from commit f10c1654fe13d797d2fd347dc47f72f93c58cf62)
---
 sysdeps/s390/fpu/s_llrint.c  | 50 +++++++++++++++++++++++++++++++++++
 sysdeps/s390/fpu/s_llrintf.c | 50 +++++++++++++++++++++++++++++++++++
 sysdeps/s390/fpu/s_llrintl.c | 51 ++++++++++++++++++++++++++++++++++++
 3 files changed, 151 insertions(+)
 create mode 100644 sysdeps/s390/fpu/s_llrint.c
 create mode 100644 sysdeps/s390/fpu/s_llrintf.c
 create mode 100644 sysdeps/s390/fpu/s_llrintl.c

diff --git a/sysdeps/s390/fpu/s_llrint.c b/sysdeps/s390/fpu/s_llrint.c
new file mode 100644
index 0000000000..edd796ae8c
--- /dev/null
+++ b/sysdeps/s390/fpu/s_llrint.c
@@ -0,0 +1,50 @@
+/* llrint() - S390 version.
+   Copyright (C) 2019 Free Software Foundation, Inc.
+
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public License as
+   published by the Free Software Foundation; either version 2.1 of the
+   License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#if defined __s390x__ && defined HAVE_S390_MIN_Z196_ZARCH_ASM_SUPPORT
+/* We only support s390x as on s390 a long long int refers to a register pair
+   of two 4byte registers instead of a 8byte register which is produced by the
+   instruction.
+   Note: On s390 this instruction would only be used if build with -mzarch.  */
+# include <math.h>
+# include <libm-alias-double.h>
+
+long long int
+__llrint (double x)
+{
+  long long int y;
+  /* The z196 zarch "convert to fixed" (cgdbra) instruction is rounding
+     according to current rounding mode (M3-field: 0).
+     First convert x with suppressed inexact exception and check if the
+     resulting value is beyond the target limits (indicated by cc=3;
+     Note: a nan is also indicated by cc=3).
+     If the resulting value is within the target limits, redo
+     without suppressing the inexact exception.  */
+  __asm__ ("cgdbra %0,0,%1,4 \n\t"
+	   "jo 1f \n\t"
+	   "cgdbra %0,0,%1,0 \n\t"
+	   "1:"
+	   : "=&d" (y) : "f" (x) : "cc");
+  return y;
+}
+libm_alias_double (__llrint, llrint)
+
+#else
+# include <sysdeps/ieee754/dbl-64/s_llrint.c>
+#endif
diff --git a/sysdeps/s390/fpu/s_llrintf.c b/sysdeps/s390/fpu/s_llrintf.c
new file mode 100644
index 0000000000..3cbe7c581a
--- /dev/null
+++ b/sysdeps/s390/fpu/s_llrintf.c
@@ -0,0 +1,50 @@
+/* llrintf() - S390 version.
+   Copyright (C) 2019 Free Software Foundation, Inc.
+
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public License as
+   published by the Free Software Foundation; either version 2.1 of the
+   License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#if defined __s390x__ && defined HAVE_S390_MIN_Z196_ZARCH_ASM_SUPPORT
+/* We only support s390x as on s390 a long long int refers to a register pair
+   of two 4byte registers instead of a 8byte register which is produced by the
+   instruction.
+   Note: On s390 this instruction would only be used if build with -mzarch.  */
+# include <math.h>
+# include <libm-alias-float.h>
+
+long long int
+__llrintf (float x)
+{
+  long long int y;
+  /* The z196 zarch "convert to fixed" (cgebra) instruction is rounding
+     according to current rounding mode (M3-field: 0).
+     First convert x with suppressed inexact exception and check if the
+     resulting value is beyond the target limits (indicated by cc=3;
+     Note: a nan is also indicated by cc=3).
+     If the resulting value is within the target limits, redo
+     without suppressing the inexact exception.  */
+  __asm__ ("cgebra %0,0,%1,4 \n\t"
+	   "jo 1f \n\t"
+	   "cgebra %0,0,%1,0 \n\t"
+	   "1:"
+	   : "=&d" (y) : "f" (x) : "cc");
+  return y;
+}
+libm_alias_float (__llrint, llrint)
+
+#else
+# include <sysdeps/ieee754/flt-32/s_llrintf.c>
+#endif
diff --git a/sysdeps/s390/fpu/s_llrintl.c b/sysdeps/s390/fpu/s_llrintl.c
new file mode 100644
index 0000000000..37eea5914f
--- /dev/null
+++ b/sysdeps/s390/fpu/s_llrintl.c
@@ -0,0 +1,51 @@
+/* llrintl() - S390 version.
+   Copyright (C) 2019 Free Software Foundation, Inc.
+
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public License as
+   published by the Free Software Foundation; either version 2.1 of the
+   License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#if defined __s390x__ && defined HAVE_S390_MIN_Z196_ZARCH_ASM_SUPPORT
+/* We only support s390x as on s390 a long long int refers to a register pair
+   of two 4byte registers instead of a 8byte register which is produced by the
+   instruction.
+   Note: On s390 this instruction would only be used if build with -mzarch.  */
+# include <math.h>
+# include <math_private.h>
+# include <libm-alias-ldouble.h>
+
+long long int
+__llrintl (_Float128 x)
+{
+  long long int y;
+  /* The z196 zarch "convert to fixed" (cgxbra) instruction is rounding
+     according to current rounding mode (M3-field: 0).
+     First convert x with suppressed inexact exception and check if the
+     resulting value is beyond the target limits (indicated by cc=3;
+     Note: a nan is also indicated by cc=3).
+     If the resulting value is within the target limits, redo
+     without suppressing the inexact exception.  */
+  __asm__ ("cgxbra %0,0,%1,4 \n\t"
+	   "jo 1f \n\t"
+	   "cgxbra %0,0,%1,0 \n\t"
+	   "1:"
+	   : "=&d" (y) : "f" (x) : "cc");
+  return y;
+}
+libm_alias_ldouble (__llrint, llrint)
+
+#else
+# include <sysdeps/ieee754/ldbl-128/s_llrintl.c>
+#endif
-- 
2.18.2

